src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```diff
@@ -49,82 +51,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
-                            $cartRules[$orderInvoice->id]['value_tax_incl'] = Tools::ps_round(
-                                $orderInvoice->total_paid_tax_incl * $discountValue / 100,
-                                2
-                            );
-                            $cartRules[$orderInvoice->id]['value_tax_excl'] = Tools::ps_round(
-                                $orderInvoice->total_paid_tax_excl * $discountValue / 100,
-                                2
-                            );
-
-                            // Update OrderInvoice
-                            $this->applyDiscountOnInvoice(
-                                $orderInvoice,
-                                $cartRules[$orderInvoice->id]['value_tax_incl'],
-                                $cartRules[$orderInvoice->id]['value_tax_excl']
-                            );
-                        }
-                    } else {
-                        $cartRules[0]['value_tax_incl'] = Tools::ps_round(
-                            $order->total_paid_tax_incl * $discountValue / 100,
-                            2
-                        );
-                        $cartRules[0]['value_tax_excl'] = Tools::ps_round(
-                            $order->total_paid_tax_excl * $discountValue / 100,
-                            2
-                        );
-                    }
-                } else {
-                    throw new OrderException('Percentage discount value cannot be higher than 100%.');
-                }
+                $this->buildPercentTypeCartRules($cartRules, $order, $discountValue, $orderInvoice);
 
                 break;
-            // Amount type
             case OrderDiscountType::DISCOUNT_AMOUNT:
-                if (isset($orderInvoice)) {
-                    if ($discountValue > $orderInvoice->total_paid_tax_incl) {
-                        throw new OrderException('The discount value is greater than the order invoice total.');
-                    }
+                $this->buildAmountTypeCartRules($cartRules, $order, $discountValue, $orderInvoice);
+
+                break;
+            case OrderDiscountType::FREE_SHIPPING:
+                $this->buildFreeShippingTypeCartRules($cartRules, $order, $orderInvoice);
+
+                break;
+            default:
+                throw new OrderException('The discount type is invalid.');
+        }
 
-                    $cartRules[$orderInvoice->id]['value_tax_incl'] = Tools::ps_round($discountValue, 2);
+        return $cartRules;
+    }
+
+    /**
+     * @param array $cartRules
+     * @param Order $order
+     * @param float $discountValue
+     * @param OrderInvoice|null $orderInvoice
+     */
+    private function buildPercentTypeCartRules(array &$cartRules, Order $order, float $discountValue, ?OrderInvoice $orderInvoice): void
+    {
+        if ($discountValue < PercentageDiscount::MAX_PERCENTAGE) {
+            if (isset($orderInvoice)) {
```

Comment:
If you only deal with an array of OrderInvoice this method can be simplified (the other two as well) you wouldn't need to deal with three different cases each time only:
```php
if (count($orderInvoices) > 0) {
 ...
} else {
 ...
}
```